walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,136 @@
+            // before checking the event availability, the entire day is available
+            // return timeslots required attendees are available for
+            requiredTimes = determineAvailableTimes(events, request, request.getAttendees());
+            // return timeslots optional attendees are available for
+            optionalTimes = determineAvailableTimes(events, request, request.getOptionalAttendees());
+        }
+        // sort times that required attendees are available for
+        List<TimeRange> finalRequiredTimes = new ArrayList<>();
+        finalRequiredTimes.addAll(requiredTimes);
+        Collections.sort(finalRequiredTimes, TimeRange.ORDER_BY_START);
+        // sort times that optional attendees are available for
+        List<TimeRange> finalOptionalTimes = new ArrayList<>();
+        finalOptionalTimes.addAll(optionalTimes);
+        Collections.sort(finalOptionalTimes, TimeRange.ORDER_BY_START);
+
+        // check to see if the lists need to be compared; if there are no attendees at all, the entire 
+        // day is available. Otherwise, if there are no optional attendees, we can
+        // return times that work for required attendees, and vice versa.
+        if (request.getAttendees().size() == 0 && request.getOptionalAttendees().size() == 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        } else if (request.getAttendees().size() == 0) {
+            return finalOptionalTimes;
+        } else if (request.getOptionalAttendees().size() == 0) {
+            return finalRequiredTimes;
+        }
+
+        // iterate through the optional attendees
+        List<TimeRange> possibleFinalTimes = new ArrayList<>();
+        possibleFinalTimes = crossCheck(finalOptionalTimes, finalRequiredTimes, request);
+
+        if (possibleFinalTimes.size() != 0) {
+            return possibleFinalTimes;
+        }
+        return finalRequiredTimes;
+    }
+
+    /**
+    * Determines which times throughout the day are available for a meeting, given events already 
+    * happening, a request, a list of attendees (comes from the request, but is inputted separately 
+    * to allow functionality for both required and optional attendees
+    */
+    public Set<TimeRange> determineAvailableTimes(Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+        Set<TimeRange> availableTimes = new HashSet<>();
+        availableTimes.add(TimeRange.WHOLE_DAY);
+        for (String person : attendees) {
+            // iterate through the events already scheduled for that day
+            for (Event mtg : events) {
+                Set<String> curAttending = mtg.getAttendees();
+                TimeRange curWhen = mtg.getWhen();
+                if (curAttending.contains(person)) {
+                    List<TimeRange> times = new ArrayList<>();
+                    times.addAll(availableTimes);
+                    // iterate through the available times
```

Comment:
One insight here is that we can avoid iterating through all of the candidate meeting slots multiple times. For example, a meeting starting at 5pm doesn't need to update availability that ends at 2pm. Another insight is that if there's a meeting that starts at 8am and ends at 9am, then 9am is the next candidate start time. If we sort meetings by start time, then this helps with both of these ideas.